File: game-development/rendering-pipeline.py
"""
Rendering Pipeline - Python

Graphics rendering concepts:
- 2D rendering
- Sprite batching
- Camera/viewport
- Z-ordering
"""

from typing import List, Dict, Tuple


# 2D Renderer
class Renderer2D:
    def __init__(self, width: int, height: int):
        self.width = width
        self.height = height
        self.sprites: List['Sprite'] = []
        self.camera = {'x': 0, 'y': 0, 'zoom': 1}
    
    def add_sprite(self, sprite: 'Sprite'):
        self.sprites.append(sprite)
    
    def sort_by_depth(self):
        self.sprites.sort(key=lambda s: s.z or 0)
    
    def render(self) -> List[Dict]:
        self.sort_by_depth()
        rendered = []
        
        for sprite in self.sprites:
            # Apply camera transform
            screen_x = (sprite.x - self.camera['x']) * self.camera['zoom']
            screen_y = (sprite.y - self.camera['y']) * self.camera['zoom']
            
            # Culling - only render if in viewport
            if self.is_in_viewport(screen_x, screen_y, sprite.width, sprite.height):
                rendered.append({
                    **sprite.__dict__,
                    'screen_x': screen_x,
                    'screen_y': screen_y
                })
        
        return rendered
    
    def is_in_viewport(self, x: float, y: float, width: float, height: float) -> bool:
        return not (x + width < 0 or x > self.width or y + height < 0 or y > self.height)
    
    def set_camera(self, x: float, y: float, zoom: float = 1):
        self.camera = {'x': x, 'y': y, 'zoom': zoom}


# Sprite
class Sprite:
    def __init__(self, x: float, y: float, width: float, height: float, texture: str, z: int = 0):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.texture = texture
        self.z = z


# Sprite Batch (for efficient rendering)
class SpriteBatch:
    def __init__(self):
        self.sprites: List[Sprite] = []
    
    def add(self, sprite: Sprite):
        self.sprites.append(sprite)
    
    def clear(self):
        self.sprites = []
    
    def render(self) -> Dict[str, List[Sprite]]:
        batches = {}
        
        for sprite in self.sprites:
            if sprite.texture not in batches:
                batches[sprite.texture] = []
            batches[sprite.texture].append(sprite)
        
        return batches


# Example usage
if __name__ == "__main__":
    print("=== Rendering Pipeline ===\n")
    
    renderer = Renderer2D(800, 600)
    
    # Add sprites
    renderer.add_sprite(Sprite(100, 100, 32, 32, 'player.png', 1))
    renderer.add_sprite(Sprite(200, 150, 32, 32, 'enemy.png', 0))
    renderer.add_sprite(Sprite(150, 120, 64, 64, 'background.png', -1))
    renderer.add_sprite(Sprite(300, 200, 32, 32, 'player.png', 1))
    
    print("Rendering with default camera:")
    rendered = renderer.render()
    for sprite in rendered:
        print(f"  {sprite['texture']} at screen ({sprite['screen_x']}, {sprite['screen_y']}), z: {sprite['z']}")
    
    print("\nRendering with camera offset:")
    renderer.set_camera(50, 50)
    rendered2 = renderer.render()
    for sprite in rendered2:
        print(f"  {sprite['texture']} at screen ({sprite['screen_x']}, {sprite['screen_y']})")
    
    # Sprite batching
    print("\nSprite Batching:")
    batch = SpriteBatch()
    batch.add(Sprite(0, 0, 32, 32, 'tile.png'))
    batch.add(Sprite(32, 0, 32, 32, 'tile.png'))
    batch.add(Sprite(64, 0, 32, 32, 'tile.png'))
    batch.add(Sprite(0, 32, 32, 32, 'grass.png'))
    
    batches = batch.render()
    for texture, sprites in batches.items():
        print(f"  Texture: {texture}, Sprites: {len(sprites)}")

File: game-development/rendering-pipeline.js
/**
 * Rendering Pipeline - JavaScript
 * 
 * Graphics rendering concepts:
 * - 2D rendering
 * - Sprite batching
 * - Camera/viewport
 * - Z-ordering
 */

// 2D Renderer
class Renderer2D {
    constructor(width, height) {
        this.width = width;
        this.height = height;
        this.sprites = [];
        this.camera = { x: 0, y: 0, zoom: 1 };
    }
    
    addSprite(sprite) {
        this.sprites.push(sprite);
    }
    
    // Sort by z-order (depth)
    sortByDepth() {
        this.sprites.sort((a, b) => (a.z || 0) - (b.z || 0));
    }
    
    // Render all sprites
    render() {
        this.sortByDepth();
        const rendered = [];
        
        for (const sprite of this.sprites) {
            // Apply camera transform
            const screenX = (sprite.x - this.camera.x) * this.camera.zoom;
            const screenY = (sprite.y - this.camera.y) * this.camera.zoom;
            
            // Culling - only render if in viewport
            if (this.isInViewport(screenX, screenY, sprite.width, sprite.height)) {
                rendered.push({
                    ...sprite,
                    screenX,
                    screenY
                });
            }
        }
        
        return rendered;
    }
    
    isInViewport(x, y, width, height) {
        return !(x + width < 0 || x > this.width || y + height < 0 || y > this.height);
    }
    
    setCamera(x, y, zoom = 1) {
        this.camera = { x, y, zoom };
    }
}

// Sprite
class Sprite {
    constructor(x, y, width, height, texture, z = 0) {
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
        this.texture = texture;
        this.z = z;
    }
}

// Sprite Batch (for efficient rendering)
class SpriteBatch {
    constructor() {
        this.sprites = [];
    }
    
    add(sprite) {
        this.sprites.push(sprite);
    }
    
    clear() {
        this.sprites = [];
    }
    
    // Batch render (group by texture for efficiency)
    render() {
        const batches = {};
        
        for (const sprite of this.sprites) {
            if (!batches[sprite.texture]) {
                batches[sprite.texture] = [];
            }
            batches[sprite.texture].push(sprite);
        }
        
        return batches;
    }
}

// Example usage
console.log('=== Rendering Pipeline ===\n');

const renderer = new Renderer2D(800, 600);

// Add sprites
renderer.addSprite(new Sprite(100, 100, 32, 32, 'player.png', 1));
renderer.addSprite(new Sprite(200, 150, 32, 32, 'enemy.png', 0));
renderer.addSprite(new Sprite(150, 120, 64, 64, 'background.png', -1));
renderer.addSprite(new Sprite(300, 200, 32, 32, 'player.png', 1));

console.log('Rendering with default camera:');
const rendered = renderer.render();
rendered.forEach(sprite => {
    console.log(`  ${sprite.texture} at screen (${sprite.screenX}, ${sprite.screenY}), z: ${sprite.z}`);
});

console.log('\nRendering with camera offset:');
renderer.setCamera(50, 50);
const rendered2 = renderer.render();
rendered2.forEach(sprite => {
    console.log(`  ${sprite.texture} at screen (${sprite.screenX}, ${sprite.screenY})`);
});

// Sprite batching
console.log('\nSprite Batching:');
const batch = new SpriteBatch();
batch.add(new Sprite(0, 0, 32, 32, 'tile.png'));
batch.add(new Sprite(32, 0, 32, 32, 'tile.png'));
batch.add(new Sprite(64, 0, 32, 32, 'tile.png'));
batch.add(new Sprite(0, 32, 32, 32, 'grass.png'));

const batches = batch.render();
for (const [texture, sprites] of Object.entries(batches)) {
    console.log(`  Texture: ${texture}, Sprites: ${sprites.length}`);
}

module.exports = { Renderer2D, Sprite, SpriteBatch };

